app/src/main/java/org/oppia/android/app/home/HomeViewModel.kt

```diff
@@ -0,0 +1,164 @@
+    // This will block until all data providers return initial results (which may be default
+    // instances). If any of the data providers are pending or failed, the combined result will also
+    // be pending or failed.
+    profileDataProvider.combineWith(
+      ongoingStoryListSummaryDataProvider,
+      PROFILE_AND_ONGOING_STORY_COMBINED_PROVIDER_ID
+    ) { profile, ongoingStoryList ->
+      listOfNotNull(
+        computeWelcomeViewModel(profile),
+        computePromotedStoryListViewModel(ongoingStoryList)
+      )
+    }.combineWith(
+      topicListSummaryDataProvider,
+      HOME_FRAGMENT_COMBINED_PROVIDER_ID
+    ) { homeItemViewModelList, topicList ->
+      homeItemViewModelList + listOf(AllTopicsViewModel()) + computeTopicSummaryItemViewModelList(
+        topicList
+      )
+    }
+  }
+
+  // Resulting LiveData to bind to the outer RecyclerView & that contains all ViewModels.
+  val homeItemViewModelListLiveData: LiveData<List<HomeItemViewModel>> by lazy {
+    Transformations.map(homeItemViewModelListDataProvider.toLiveData()) { itemListResult ->
+      if (itemListResult.isFailure()) {
+        logger.e(
+          "HomeFragment",
+          "Failed to retrieve fragment",
+          itemListResult.getErrorOrNull()
+        )
+      }
+      return@map itemListResult.getOrDefault(listOf())
+    }
+  }
+
+  private fun computeWelcomeViewModel(profile: Profile): HomeItemViewModel? {
+    return if (profile.name.isNotEmpty()) {
+      WelcomeViewModel(fragment, oppiaClock, profile.name)
+    } else null
+  }
+
+  private fun computePromotedStoryListViewModel(
+    ongoingStoryList: OngoingStoryList
+  ): HomeItemViewModel? {
+    val storyViewModelList = computePromotedStoryViewModelList(ongoingStoryList)
+    return if (storyViewModelList.isNotEmpty()) {
+      return PromotedStoryListViewModel(
+        activity,
+        internalProfileId,
+        intentFactoryShim,
+        storyViewModelList
+      )
+    } else null
+  }
+
+  private fun computePromotedStoryViewModelList(
+    ongoingStoryList: OngoingStoryList
+  ): List<PromotedStoryViewModel> {
+    if (ongoingStoryList.recentStoryCount != 0) {
+      return ongoingStoryList.recentStoryList.take(limit)
+        .map { promotedStory ->
+          PromotedStoryViewModel(
+            activity,
+            internalProfileId,
+            intentFactoryShim,
+            ongoingStoryList.recentStoryCount,
+            storyEntityType,
+            promotedStory
+          )
+        }
+    } else {
+      // TODO(#936): Optimise this as part of recommended stories.
+      return ongoingStoryList.olderStoryList.take(limit)
+        .map { promotedStory ->
+          PromotedStoryViewModel(
+            activity,
+            internalProfileId,
+            intentFactoryShim,
+            ongoingStoryList.olderStoryCount,
+            storyEntityType,
+            promotedStory
+          )
+        }
+    }
```

Comment:
```suggestion
    val storyList = if (ongoingStoryList.recentStoryCount != 0) {
      ongoingStoryList.recentStoryList
    } else {
      // TODO(#936): Optimise this as part of recommended stories.
      ongoingStoryList.olderStoryList
    }
    return storyList.take(limit)
      .map { promotedStory ->
        PromotedStoryViewModel(
          activity,
          internalProfileId,
          intentFactoryShim,
          storyList.size,
          storyEntityType,
          promotedStory
        )
      }
```

It appears that this can be simplified.

Reply:
Done -- thanks!